mesonbuild/compilers/compilers.py

```diff
@@ -462,9 +467,50 @@ def __init__(self, compiler: T.Union['Compiler', StaticLinker],
                  iterable: T.Optional[T.Iterable[str]] = None):
         self.compiler = compiler
         self.__container = list(iterable) if iterable is not None else []  # type: T.List[str]
-        self.__seen_args = set()
-        for arg in self.__container:
-            self.__seen_args.add(arg)
+        self.pre = deque()
+        self.post = deque()
+
+    # Flush the saved pre and post list into the __container list
+    #
+    # This correctly deduplicates the entries after _can_dedup definition
+    # Note: This function is designed to work without delete operations, as deletions are worsening the performance a lot.
+    def flush_pre_post(self):
+        pre_flush = deque()
+        pre_flush_set = set()
+        post_flush = deque()
+        post_flush_set = set()
+
+        #The two lists are here walked from the front to the back, in order to not need removals for deduplication
+        self.pre.reverse()
+        for a in self.pre:
+            dedup = self._can_dedup(a)
+            if a not in pre_flush_set:
+                pre_flush.appendleft(a)
+                if dedup == 2:
+                    pre_flush_set.add(a)
+
+        self.post.reverse()
+        for a in self.post:
+            dedup = self._can_dedup(a)
+            if a not in post_flush_set:
+                post_flush.appendleft(a)
+                if dedup == 2:
+                    post_flush_set.add(a)
+
+        #pre and post will overwrite every element that is in the container
+        #only copy over args that are in __container but not in the post flush or pre flush set
+
+        for a in self.__container:
+            if a not in post_flush_set and a not in pre_flush_set:
+                pre_flush.append(a)
+
+        self.__container = list(pre_flush) + list(post_flush)
+        self.pre.clear()
+        self.post.clear()
+
+    def __iter__(self):
+        self.flush_pre_post()
+        return self.__container.__iter__();
```

Comment:
```suggestion
        return iter(self.__container)
```

Don't call dunder methods directly (that's what the methods that start and end with `__` are called) unless you're doing something tricky with subclassing. Use the builtin function instead.